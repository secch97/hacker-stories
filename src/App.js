import { useEffect, useReducer, useRef, useState } from "react";

const useSemiPersistentState = (key, initialState) => {
  const [value, setValue] = useState(localStorage.getItem(key) ?? initialState);

  useEffect(() => {
    localStorage.setItem(key, value.trim());
  }, [value, key]);

  return [value, setValue];
};

const storiesReducer = (state, action) => {
  switch(action.type) {
    case "STORIES_FETCH_INIT":
      return{
        ...state,
        isLoading: true,
        isError:false
      }
    case "STORIES_FETCH_SUCCESS":
      return {
        ...state,
        isLoading: false,
        isError: false,
        data: action.payload
      };
    case "STORIES_FETCH_FAILURE":
      return {
        ...state,
        isLoading: false,
        isError: true
      }
    case "REMOVE_STORY":
      return {
        ...state,
        data: state.data.filter(
          (story) => action.payload !== story.objectID
        )
      };
    default:
      throw new Error();
  };
};

const App = () => {

  const initialStories = [
    {
      title: "React",
      url: "https://reactjs.org/",
      author: "Jordan Walke",
      num_comments: 3,
      points: 4,
      objectID: 0,
    },
    {
      title: "Redux",
      url: "https://redux.js.org/",
      author: "Dan Abramov, Andrew Clark",
      num_comments: 2,
      points: 5,
      objectID: 1,
    },
    {
      title: "Road to React",
      url: "https://www.roadtoreact.com/",
      author: "Robin Wieruch",
      num_comments: 10,
      points: 15,
      objectID: 2,
    },
  ];

  const getAsyncStories = () => {
    return new Promise((resolve) => {
      setTimeout(() => {
        resolve({
          data: {
            stories: initialStories,
          },
        });
      }, 2000);
    });
  };

  /* 
    ======================
    =       HOOKS        =
    ======================     
  */
  const [searchTerm, setSearchTerm] = useSemiPersistentState("search", "React");
  const [stories, dispatchStories] = useReducer(storiesReducer, {
      data: [],
      isLoading: false,
      isError: false
    });

  useEffect(() => {
    dispatchStories({type: "STORIES_FETCH_INIT"});

    getAsyncStories().then((result) => {
      dispatchStories({
      type: "STORIES_FETCH_SUCCESS",
      payload: result.data.stories
      });
    }).catch(() => {
      dispatchStories({type: "STORIES_FETCH_FAILURE"});
    });
  }, []);

  /* 
    ======================
    =      HANDLERS      =
    ======================     
  */
  const handleSearch = (event) => {
    const newSearchTerm = event.target.value;
    setSearchTerm(newSearchTerm);
  };

  const handleRemoveStory = (objectID) => {
    dispatchStories({
      type: "REMOVE_STORY",
      payload: objectID
    });  
  };

  /* 
    ======================
    =       HELPERS      =
    ======================     
  */
  const searchedStories = stories.data.filter((story) =>
    story.title.toLowerCase().includes(searchTerm.toLowerCase().trim())
  );

  return (
    <>
      <h1>My Hacker Stories</h1>
      <InputWithLabel
        id="search"
        value={searchTerm}
        onInputChange={handleSearch}
        isFocused
      >
        <strong>Search:</strong>
      </InputWithLabel>
      <hr />
      {stories.isError && (<p>Something went wrong...</p>)}
      {
        stories.isLoading ? 
        (<p>Loading...</p>) : 
        (<List list={searchedStories} onRemoveItem={handleRemoveStory} />)
      }

    </>
  );
};

const InputWithLabel = ({
  id,
  value,
  type = "text",
  onInputChange,
  isFocused,
  children,
}) => {
  //A
  const inputRef = useRef();

  useEffect(() => {
    if (isFocused && inputRef.current) {
      inputRef.current.focus();
    }
  }, [isFocused]);

  return (
    <div>
      <label htmlFor={id}>{children}</label>
      &nbsp;
      <input
        ref={inputRef}
        id={id}
        type={type}
        value={value}
        onChange={onInputChange}
      />
    </div>
  );
};

const List = ({ list, onRemoveItem }) => {
  return (
    <ul>
      {list.map((listItem) => (
        <Item
          key={listItem.objectID}
          {...listItem}
          onRemoveItem={onRemoveItem}
        />
      ))}
    </ul>
  );
};

const Item = ({
  objectID,
  title,
  url,
  author,
  num_comments,
  points,
  onRemoveItem,
}) => {
  return (
    <li>
      <span>
        <a href={url}>{title}</a>
      </span>
      <span> {author}</span>
      <span> {num_comments}</span>
      <span> {points}</span>
      <span>
        <button type="button" onClick={() => onRemoveItem(objectID)}>
          Dismiss
        </button>
      </span>
    </li>
  );
};

export { App as default };
