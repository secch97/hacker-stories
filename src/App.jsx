import { useCallback, useEffect, useReducer, useState } from "react";
import axios from "axios";
import styled from "styled-components";
import SearchForm from "./SearchForm";
import List from "./List";

const StyledContainer = styled.div`
  height: 100vw;
  padding: 20px;
  background: #83a4d4;
  background: linear-gradient(to left, #b6fbff, #83a4d4);
  color: #171212;
`;

const StyledHeadlinePrimary = styled.h1`
  font-size: 48px;
  font-weight: 300px;
  letter-spacing: 2px;
`;

const useSemiPersistentState = (key, initialState) => {
  const [value, setValue] = useState(localStorage.getItem(key) ?? initialState);

  useEffect(() => {
    localStorage.setItem(key, value.trim());
  }, [value, key]);

  return [value, setValue];
};

const storiesReducer = (state, action) => {
  switch (action.type) {
    case "STORIES_FETCH_INIT":
      return {
        ...state,
        isLoading: true,
        isError: false,
      };
    case "STORIES_FETCH_SUCCESS":
      return {
        ...state,
        isLoading: false,
        isError: false,
        data: action.payload,
      };
    case "STORIES_FETCH_FAILURE":
      return {
        ...state,
        isLoading: false,
        isError: true,
      };
    case "REMOVE_STORY":
      return {
        ...state,
        data: state.data.filter((story) => action.payload !== story.objectID),
      };
    default:
      throw new Error();
  }
};

//A
const API_ENDPOINT = "https://hn.algolia.com/api/v1/search?query=";

const App = () => {
  /* 
    ======================
    =       HOOKS        =
    ======================     
  */
  const [searchTerm, setSearchTerm] = useSemiPersistentState("search", "React");
  const [url, setUrl] = useState(`${API_ENDPOINT}${searchTerm}`);
  const [stories, dispatchStories] = useReducer(storiesReducer, {
    data: [],
    isLoading: false,
    isError: false,
  });

  const handleFetchStories = useCallback(async () => {
    dispatchStories({ type: "STORIES_FETCH_INIT" });
    try {
      const result = await axios.get(url);
      dispatchStories({
        type: "STORIES_FETCH_SUCCESS",
        payload: result.data.hits,
      });
    } catch (error) {
      dispatchStories({ type: "STORIES_FETCH_FAILURE" });
    }
  }, [url]);

  useEffect(() => {
    handleFetchStories();
  }, [handleFetchStories]);

  /* 
    ======================
    =      HANDLERS      =
    ======================     
  */

  const handleSearchInput = (event) => {
    const newSearchTerm = event.target.value;
    setSearchTerm(newSearchTerm);
  };

  const handleSearchSubmit = (event) => {
    setUrl(`${API_ENDPOINT}${searchTerm}`);
    event.preventDefault();
  };

  const handleRemoveStory = (objectID) => {
    dispatchStories({
      type: "REMOVE_STORY",
      payload: objectID,
    });
  };

  return (
    <StyledContainer>
      <StyledHeadlinePrimary>My Hacker Stories</StyledHeadlinePrimary>
      <SearchForm
        searchTerm={searchTerm}
        onSearchInput={handleSearchInput}
        onSearchSubmit={handleSearchSubmit}
      />
      {stories.isError && <p>Something went wrong...</p>}
      {stories.isLoading ? (
        <p>Loading...</p>
      ) : (
        <List list={stories.data} onRemoveItem={handleRemoveStory} />
      )}
    </StyledContainer>
  );
};

export { App as default };
