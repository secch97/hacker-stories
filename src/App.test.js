/* eslint-disable testing-library/no-debugging-utils */
import React from 'react';

import App, {
  storiesReducer, 
  Item, 
  List, 
  SearchForm, 
  InputWithLabel
} from './App';

import {
  render,
  screen,
  fireEvent,
  act
} from "@testing-library/react"

import axios from 'axios';

/*
  Before we unit test our first React component, we will cover how to test
  just a JS function, in this case, storiesReducer and one of its actions.
*/

const storyOne = {
  title: "React",
  url: "https://reactjs.org",
  author: "Jordan Walke",
  num_comments: 3,
  points: 4,
  objectID: 0
};


const storyTwo = {
  title: "Redux",
  url: "https://redux.js.org",
  author: "Dan Abramov, Andrew Clark",
  num_comments: 2,
  points: 5,
  objectID: 1
};

const stories = [storyOne, storyTwo];

describe('storiesReducer', () => { 
  test("Removes a story from all stories", ()=>{

    const action = {
      type: "REMOVE_STORY",
      payload: storyOne.objectID
    };
    
    const state = {
      data: stories,
      isLoading: false,
      isError: false
    };

    const newState = storiesReducer(state, action);

    const expectedState = {
      data: [storyTwo],
      isLoading: false, 
      isError: false
    };

    expect(newState).toStrictEqual(expectedState);
  });

  test('Stories fetch init', () => { 
    const state = {
      data: [],
      isLoading: false,
      isError: false
    };

    const action = {
      type: "STORIES_FETCH_INIT",
    };

    const newState = storiesReducer(state, action);
    const expectedState = {
      data: [],
      isLoading: true,
      isError: false
    }

    expect(newState).toStrictEqual(expectedState);
  });

  test('Stories fetch success', async () => { 
      const state = {
        data: [],
        isLoading: true,
        isError: false
      };

      const result = await axios.get("https://hn.algolia.com/api/v1/search?query=React");
      const action = {
        type: "STORIES_FETCH_SUCCESS",
        payload: result.data.hits
      }

      const newState = storiesReducer(state, action);

      const expectedState = {
        data: result.data.hits,
        isLoading: false,
        isError: false
      };
      expect(newState).toStrictEqual(expectedState);
   });
 });

 describe('Item component', () => { 
    test('Should render item component properties', () => { 
      render(<Item {...storyOne}/>);
      screen.debug();
      expect(screen.getByText("Jordan Walke")).toBeTruthy();
      expect(screen.getByText("React")).toHaveAttribute("href", "https://reactjs.org");
     });

     test('Renders a clickable dismiss button', () => { 
        render(<Item {...storyOne}></Item>);
        expect(screen.getByRole("button")).toBeInTheDocument();
      });

      test('Clicking the dismiss button calls the callback handler', () => { 
          const handleRemoveItem = jest.fn();

          render(<Item {...storyOne} onRemoveItem={handleRemoveItem}></Item>);
          fireEvent.click(screen.getByRole("button"));
          expect(handleRemoveItem).toHaveBeenCalledTimes(1);
       });
  });

  describe("SearchForm", () => {
    const searchFormProps = {
      searchTerm: "React",
      onSearchInput: jest.fn(),
      onSearchSubmit: jest.fn()
    };

    test('Renders the input field with its values', () => { 
      render(<SearchForm {...searchFormProps}/>);
      screen.debug();
      expect(screen.getByDisplayValue("React")).toBeInTheDocument();
     });

     test("Renders the correct label", () => {
      render(<SearchForm {...searchFormProps}/>);
      expect(screen.getByLabelText(/Search/)).toBeInTheDocument();
     });

     test("Calls onSearchInput on input field change", ()=>{
      render(<SearchForm {...searchFormProps}/>);
      fireEvent.change(screen.getByDisplayValue("React"), {
        target: {value: "Redux"}
      });
      expect(searchFormProps.onSearchInput).toHaveBeenCalledTimes(1);
     });

     test("Calls onSearchSubmit on button submit click", ()=>{
      render(<SearchForm {...searchFormProps}></SearchForm>);
      fireEvent.submit(screen.getByRole("button"));
      expect(searchFormProps.onSearchSubmit).toHaveBeenCalledTimes(1);
     });
  });