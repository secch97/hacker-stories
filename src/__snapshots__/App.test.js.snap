// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchForm Renders snapshot 1`] = `
<form
  class="sc-kAyceB bWZyfg"
>
  <div>
    <label
      class="sc-imWYAI klviXL label"
      for="search"
    >
      <strong>
        Search: 
      </strong>
    </label>
    Â 
    <input
      class="sc-jXbUNg kFkJny input"
      id="search"
      type="text"
      value="React"
    />
  </div>
  <button
    class="sc-dcJsrY sc-gsFSXq fjdjpV bESgpt"
    type="submit"
  >
    Submit
  </button>
</form>
`;
